src/radical/pilot/agent/executing/popen.py

```diff
@@ -369,42 +379,52 @@ def _handle_task(self, task):
         # store pid for last-effort termination
         _pids.append(task['proc'].pid)
 
-        self._watch_queue.put(task)
+        # handle task timeout if needed
+        self.handle_timeout(task)
+
+        # watch task for completion
+        self._watch_queue.put([self.TO_WATCH, task])
 
 
     # --------------------------------------------------------------------------
     #
     def _watch(self):
 
+        to_watch  = list()  # list of task dicts
+        to_cancel = list()  # list of task IDs
```

Comment:
maybe to have it as a `set`? it gives a better performance for lookup and in case of duplication requests for task cancelation